packages/core/src/validate/run.ts

```diff
@@ -46,6 +52,29 @@ interface ValidationErrorOpcode extends ValidationErrorBase {
   kind: 'delegatecall' | 'selfdestruct';
 }
 
+interface ValidationErrorUpgradeability extends ValidationErrorBase {
+  kind: 'no-public-upgrade-fn';
+}
+
+function skipCheck<T>(error: string, node: T): boolean {
+  function isAllowed(error: string, documentation: undefined | string | StructuredDocumentation): boolean {
+    switch (typeof documentation) {
+      case 'undefined':
+        return false;
+
+      case 'string': {
+        const entries = documentation.split(' ');
+        return ['@custom:openzeppelin-upgrade-allow', error].every(requierement => entries.includes(requierement));
+      }
```

Comment:
This doesn't look right!! We should only consider the arguments to a `@custom:openzeppelin-upgrade-allow` directive (not all of the words in the docstring), and we should throw an error if we don't recognize one of the arguments.

We need to properly parse the NatSpec. Look at `parseNatSpec` in solidity-docgen. It needs a small change because it's only looking for keywords that match the pattern `@\w+`, which is no longer the right pattern. See https://github.com/ethereum/solidity/issues/8802#issuecomment-788062522 for the allowed characters.

Reply:
can you explain `/^(?:@(\w+) )?((?:(?!^@\w+ )[^])*)/m` ?

Reply:
@frangio please note the pattern in your link is not the final one: https://github.com/ethereum/solidity/issues/8802#issuecomment-806707040

`@custom:([a-z][a-z-]*)` is the correct one for custom tag names.

Reply:
@axic: thanks for the update!

@frangio What do you think of:
```
function* execall(re: RegExp, text: string) {
  re = new RegExp(re, re.flags + (re.sticky ? '' : 'y'));
  while (true) {
    const match = re.exec(text);
    if (match && match[0] !== '') {
      yield match;
    } else {
      break;
    }
  }
}

function skipCheck(error: string, node: Node): boolean {
  function isAllowed(error: string, documentation: StructuredDocumentation | string): boolean {
    switch (typeof documentation) {
      case 'string': {
        for (const { groups } of execall(/^(?:@(?<title>\w+)(?::(?<tag>[a-z][a-z-]*))? )?(?<args>(?:(?!^@\w+ )[^])*)/m, documentation)) {
          if (groups && groups.title === 'custom' && groups.tag === 'openzeppelin-upgrade-allow') {
            for (const [, arg] of execall(/(?<arg>[a-z][a-z-]*)\s?/,groups.args)) {
              if (arg === error) {
                return true;
              }
              if (![
                'state-variable-assignment',
                'state-variable-immutable',
                'external-library-linking',
                'struct-definition',
                'enum-definition',
                'constructor',
                'delegatecall',
                'selfdestruct',
                'no-public-upgrade-fn',
              ].includes(arg)) {
                throw new Error(`NatSpec: openzeppelin-upgrade-allow argument not recognized: ${arg}`);
              }
            }
          }
        }
        return false;
      }
      default:
        return isAllowed(error, documentation.text);
    }
  }
  if ('documentation' in node && node.documentation) {
    return isAllowed(error, node.documentation);
  } else {
    return false;
  }
}```